File: sdks/python/examples/03_chain_audit.py
"""Iterate the chain ledger and pull an incident report.

After running the other examples to populate some chains, this script
shows how to retrieve them and inspect what fired.

Usage:
    HEIMDALL_API_KEY=hd_test_... python 03_chain_audit.py
"""
import os
import sys

from heimdall import Heimdall

# Heimdall returns reason strings that include math symbols (⊆) and unicode
# arrows. Make the console resilient on Windows / non-UTF-8 terminals.
if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")


def main() -> None:
    api_key = os.environ.get("HEIMDALL_API_KEY")
    if not api_key:
        raise SystemExit("Set HEIMDALL_API_KEY (your hd_test_... or hd_live_... key).")

    base_url = os.environ.get("HEIMDALL_URL", "http://localhost:8000")

    with Heimdall(api_key=api_key, base_url=base_url) as hd:
        chains = hd.list_chains(limit=10)
        print(f"{len(chains)} recent chain(s):\n")
        for c in chains:
            print(f"  [{c.status:>7}] {c.chain_id[:8]}  {c.hop_count} hop(s)  {c.head_caller} -> {c.head_callee}")

        if not chains:
            return

        target = next((c for c in chains if c.status == "denied"), chains[0])
        detail = hd.get_chain(target.chain_id)
        print(f"\nchain {target.chain_id[:8]} ({target.status}):")
        for hop in detail["hops"]:
            print(f"  hop: {hop.from_agent} -> {hop.to_agent}  action={hop.action}  scope={hop.scope}")
        for ev in detail["evaluations"]:
            mark = {"ALLOW": "ok ", "FLAG": "!  ", "DENY": "x  "}.get(ev["result"], "   ")
            print(f"  {mark} [{ev['layer']:>8}] {ev['rule']:<28}  {ev.get('reason', '')}")

        # If you've also POSTed /api/audit/report/{chain_id} you can pull the
        # Markdown incident report:
        report = hd.get_audit(target.chain_id)
        if report.exists:
            print(f"\nincident report (severity={report.severity}):")
            print(report.report)
        else:
            print("\nNo incident report yet for this chain.")


if __name__ == "__main__":
    main()

File: sdks/typescript/examples/03_chain_audit.ts
/**
 * Iterate the chain ledger and pull an incident report.
 *
 * After running the other examples to populate some chains, this script
 * shows how to retrieve them and inspect what fired.
 *
 * Usage:
 *   HEIMDALL_API_KEY=hd_test_... npx tsx examples/03_chain_audit.ts
 */
import { Heimdall, hasReport } from "../src/index";

async function main(): Promise<void> {
  const apiKey = process.env.HEIMDALL_API_KEY;
  if (!apiKey) {
    throw new Error("Set HEIMDALL_API_KEY (your hd_test_... or hd_live_... key).");
  }
  const baseUrl = process.env.HEIMDALL_URL ?? "http://localhost:8000";

  const hd = new Heimdall({ apiKey, baseUrl });

  const chains = await hd.listChains({ limit: 10 });
  console.log(`${chains.length} recent chain(s):\n`);
  for (const c of chains) {
    const status = c.status.padStart(7);
    console.log(
      `  [${status}] ${c.chain_id.slice(0, 8)}  ${c.hop_count} hop(s)  ${c.head_caller} -> ${c.head_callee}`,
    );
  }
  if (chains.length === 0) return;

  const target = chains.find((c) => c.status === "denied") ?? chains[0]!;
  const detail = await hd.getChain(target.chain_id);
  console.log(`\nchain ${target.chain_id.slice(0, 8)} (${target.status}):`);
  for (const hop of detail.hops) {
    console.log(
      `  hop: ${hop.from_agent} -> ${hop.to_agent}  action=${hop.action}  scope=${JSON.stringify(hop.scope)}`,
    );
  }
  for (const ev of detail.evaluations) {
    const mark = ev.result === "ALLOW" ? "ok " : ev.result === "FLAG" ? "!  " : "x  ";
    const layer = (ev.layer ?? "").padStart(8);
    console.log(`  ${mark} [${layer}] ${ev.rule.padEnd(28)} ${ev.reason ?? ""}`);
  }

  const report = await hd.getAudit(target.chain_id);
  if (hasReport(report)) {
    console.log(`\nincident report (severity=${report.severity}):`);
    console.log(report.report);
  } else {
    console.log("\nNo incident report yet for this chain.");
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
